lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -741,36 +750,137 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+
     if (mlir::isa<arith::AddFOp>(arithOrMathOp)) {
+      Value dstIndex = i32(0, builder);
+      builder.create<ttkernel::TileRegsAcquireOp>(arithOrMathOp.getLoc());
       builder.create<ttkernel::AddTilesOp>(arithOrMathOp.getLoc(), inCB0, inCB1,
                                            inCB0TileIndex, inCB1TileIndex,
-                                           dstTileIndex);
+                                           dstIndex);
+      builder.create<ttkernel::TileRegsCommitOp>(arithOrMathOp.getLoc());
+      builder.create<ttkernel::TileRegsWaitOp>(arithOrMathOp.getLoc());
+      builder.create<ttkernel::PackTileOp>(
+          arithOrMathOp.getLoc(), dstIndex, outCB,
+          iterators[blockArgIteratorMapping[2]]);
+      builder.create<ttkernel::TileRegsReleaseOp>(arithOrMathOp.getLoc());
     } else if (mlir::isa<arith::MulFOp>(arithOrMathOp)) {
-      builder.create<ttkernel::MulTilesOp>(arithOrMathOp.getLoc(), inCB0, inCB1,
-                                           inCB0TileIndex, inCB1TileIndex,
-                                           dstTileIndex);
+      commonComputeMulOp(arithOrMathOp, cbOperands, iterators,
+                         blockArgIteratorMapping, builder);
+    } else if (mlir::isa<arith::DivFOp>(arithOrMathOp)) {
+
+      SmallVector<std::int64_t> operandIndicesRecip;
+      // For DIV, input 1 is going through reciprocal.
+      operandIndicesRecip.push_back(1);
+      commonComputeRecipOp(arithOrMathOp, cbOperands, iterators,
+                           blockArgIteratorMapping, builder,
+                           operandIndicesRecip);
+
+      builder.create<ttkernel::MulTilesInitOp>(arithOrMathOp.getLoc(), inCB0,
+                                               inCB1);
+
+      commonComputeMulOp(arithOrMathOp, cbOperands, iterators,
+                         blockArgIteratorMapping, builder);
+
+      Value one = i32(1, builder);
+      builder.create<ttkernel::CBPopFrontOp>(arithOrMathOp.getLoc(), inCB1,
+                                             one);
     } else {
-      llvm_unreachable("Unhandled binary op compute conversion.");
+      llvm_unreachable("Unhandled conversion for operation which is neither "
+                       "unary nor binary.");
     }
   }
 
+  void commonComputeMulOp(Operation &op, ArrayRef<BlockArgument> cbOperands,
+                          ArrayRef<BlockArgument> iterators,
+                          SmallVector<unsigned> blockArgIteratorMapping,
+                          OpBuilder &builder) const {
+
+    auto inCB0 = cbOperands[0];
+    auto inCB1 = cbOperands[1];
+    auto outCB = cbOperands[2];
+
+    Value dstIndex = i32(0, builder);
+
+    builder.create<ttkernel::TileRegsAcquireOp>(op.getLoc());
+    if (mlir::isa<arith::MulFOp>(op)) {
+      builder.create<ttkernel::MulTilesOp>(
+          op.getLoc(), inCB0, inCB1, iterators[blockArgIteratorMapping[0]],
+          iterators[blockArgIteratorMapping[0]], dstIndex);
```

Comment:
We should be explicit here

```
builder.create<ttkernel::MulTilesOp>(arithOrMathOp.getLoc(), inCB0, inCB1,
                                           inCB0TileIndex, inCB1TileIndex,
                                           dstTileIndex);
```

and not rely on the fact that iterators will be the same.

Extract them into variables.